animation/zoom/zoom.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+ * animation/zoom/zoom.cpp
+ *
+ * libanimation is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * libanimation is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with eos-companion-app-service.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ *
+ * Animation that causes a surface to zoom from one rectangle to another.
+ */
+
+#include <algorithm>
+#include <array>
+
+#include <animation/box_calculation.h>
+#include <animation/geometry.h>
+#include <animation/math.h>
+#include <animation/transform_calculation.h>
+#include <animation/zoom/zoom.h>
+
+#include <glm/glm.hpp>
+#include <glm/gtc/type_ptr.hpp>
+#include <glm/gtc/matrix_transform.hpp>
+
+namespace agd = animation::geometry::dimension;
+namespace abc = animation::box_calculation;
+namespace am = animation::math;
+namespace atc = animation::transform_calculation;
+namespace az = animation::zoom;
+
+namespace
+{
+    animation::Box <animation::Point> EnsureNonZeroArea (animation::Box <animation::Point> const &box)
+    {
+        auto x1 = agd::get <0> (box.topLeft());
+        auto y1 = agd::get <1> (box.topLeft());
+        auto x2 = agd::get <0> (box.bottomRight());
+        auto y2 = agd::get <1> (box.bottomRight());
```

Comment:
This is more of a sanity check to prevent zero division in case the caller passes in a zero-area geometry.

The idea with this library was to allow external users to use it, so it should do the right thing if someone tries to scale to a zero-sized box.

Reply:
I think @joaquimrocha's point is simply about factoring out `box.topLeft()` and `box.bottomRight()` into two variables before this block. (I agree)